pkg/ring/kv/multi.go

```diff
@@ -0,0 +1,335 @@
+	err := m.runWithPrimaryClient(ctx, func(newCtx context.Context, primary kvclient) error {
+		var err2 error
+		result, err2 = primary.client.Get(newCtx, key)
+		return err2
+	})
+	return result, err
+}
+
+func (m *multiClient) CAS(ctx context.Context, key string, f func(in interface{}) (out interface{}, retry bool, err error)) error {
+	return m.runWithPrimaryClient(ctx, func(newCtx context.Context, primary kvclient) error {
+		return primary.client.CAS(newCtx, key, f)
+	})
+}
+
+func (m *multiClient) WatchKey(ctx context.Context, key string, f func(interface{}) bool) {
+	_ = m.runWithPrimaryClient(ctx, func(newCtx context.Context, primary kvclient) error {
+		primary.client.WatchKey(newCtx, key, f)
+		return newCtx.Err()
+	})
+}
+
+func (m *multiClient) WatchPrefix(ctx context.Context, prefix string, f func(string, interface{}) bool) {
+	_ = m.runWithPrimaryClient(ctx, func(newCtx context.Context, primary kvclient) error {
+		primary.client.WatchPrefix(newCtx, prefix, f)
+		return newCtx.Err()
+	})
+}
+
+func (m *multiClient) watchChanges(prefix string) {
+	for m.ctx.Err() == nil {
+		err := m.runWithPrimaryClient(m.ctx, func(newCtx context.Context, primary kvclient) error {
+			primary.client.WatchPrefix(newCtx, prefix, func(key string, val interface{}) bool {
+				level.Debug(util.Logger).Log("msg", "value updated", "key", key, "store", primary.name)
+
+				if m.mirroringEnabled.Load() {
+					// don't pass new context to keyUpdated, we don't want to react on primary client changes here.
+					m.keyUpdated(primary, key, val)
+				}
+
+				err := m.rateLimiter.Wait(newCtx)
+				if err != nil && err != context.Canceled {
+					level.Error(util.Logger).Log("msg", "error while rate limiting multi-client", "err", err)
+				}
+
+				return true
+			})
+
+			return m.ctx.Err()
+		})
+
+		if err != nil {
+			level.Warn(util.Logger).Log("msg", "watching changes returned error", "err", err)
+		}
+	}
+}
+
+func (m *multiClient) keyUpdated(primary kvclient, key string, newValue interface{}) {
+	// let's propagate this to all remaining clients
+	for _, kvc := range m.clients {
+		if kvc == primary {
+			continue
+		}
+
+		stored := true
+		err := kvc.client.CAS(m.ctx, key, func(in interface{}) (out interface{}, retry bool, err error) {
+			stored = true
+
+			if eq, ok := in.(withEqual); ok && eq.Equal(newValue) {
+				stored = false
+				level.Debug(util.Logger).Log("msg", "no change", "key", key)
+				return nil, false, nil
+			}
+
+			// try once
+			return in, false, nil
+		})
+
+		if err != nil {
+			level.Error(util.Logger).Log("msg", "failed to update value in secondary store", "key", key, "err", err, "primary", primary.name, "secondary", kvc.name)
+		} else if stored {
+			level.Info(util.Logger).Log("msg", "stored updated value to secondary store", "key", key, "primary", primary.name, "secondary", kvc.name)
+		}
+	}
+}
+
+func (m *multiClient) Stop() {
+	m.cancel()
+
+	// TODO: call Stop method of all clients from gossiping ring branch.
```

Comment:
Are you planning on fixing this in this PR? Since both of these clients are only called within the MultClient it shouldn't be a major issue but it is likely good practice.

Reply:
That TODO refers to new Stop method introduced by PR #1727. It's not part of Client interface on this branch yet.

Perhaps I should just remove Stop method here, and introduce it again when either merging this, or merging PR #1727. It kind of depends on which PR is merged first :-(

Reply:
I've renamed method to 'stop' and removed TODO. We be handled later, when merging with PR 1727. Compilation error (eg. missing Stop method when merging 1727) will make sure it's not forgotten.

Reply:
With gossiping code in master now, I've rebased and added this code.